docs/source/06_solver_hooks.rst

```diff
@@ -0,0 +1,179 @@
+
+In the :ref:`User Interface Hooks <user_interface_hooks-section>` section is explained that the plugins are allowed to
+customize the |alfasim|'s user interface extending it in order to get some input data from the user. Using the solver
+:py:func:`HOOK_INITIALIZE<alfasim_sdk.hook_specs.initialize>` the developer can obtain the input data from user interface and use it to
+initialize the plugin internal data.
+
+As already mentioned, it is allowed that the plugins have internal data, in which can hold some important information that
+will be used during the simulation, and also by other `hooks`.
+
+.. warning::
+    The plugin internal data will be hold by |alfasim|'s solver, however the plugin has full responsibility to allocate
+    and deallocate its data from memory.
+
+.. autofunction:: alfasim_sdk.hook_specs.initialize
+
+As can be seen in the example above the function :cpp:func:`set_plugin_data` is used to tell the |alfasim|'s solver to
+hold the plugin internal data.
+
+.. note::
+    Since |alfasim|'s solver uses multi-threads to perform all possible parallelizable calculation, it is important that
+    the plugins provide internal data to each `thread` to avoid data access concurrency problems. As can be seen the
+    ``HOOK_INITIALIZE`` example above, a ``for-loop`` is performed over the `threads` to set the plugin internal data.
+    The |sdk| API function :cpp:func:`get_number_of_threads` is used to do it properly. See
+    :ref:`plugin_internal_data` section for more information.
+
+.. autofunction:: alfasim_sdk.hook_specs.finalize
+
+As can be seen in the example above the function :cpp:func:`get_plugin_data` is used to retrieved the plugin internal
+data for each `thread` identified as ``thread_id``.
+
+.. note::
+    In the examples of usage of both :py:func:`HOOK_INITIALIZE<alfasim_sdk.hook_specs.initialize>` and :py:func:`HOOK_FINALIZE<alfasim_sdk.hook_specs.finalize>`
+    there are many times where an error code is returned (``return errcode;`` or ``return OK;``). As can be seen, it is
+    possible to return error codes from |sdk| API functions, however the developer can intercept this error code and
+    handle it instead of returning it to the |alfasim|'s Solver.
+
+.. _update_secondary_variables:
+
+Update plugin variables
+-----------------------
+
+The `hooks` described in this section are related to plugin secondary variables that were registered in the python config
+file, as already explained in :ref:`solver_customization` section. They are called `secondary variables` because they
+are not obtained from |alfasim|'s Solver, these ones are called primary variables and examples of those variables are `pressure`,
+`temperature`, `volume fractions` and `velocities`.
+
+Using the following `hooks` the plugin is able to calculate/update those variables in three different moments of the
+simulation step. Two of them are called in the `Hydrodynamic Solver` scope and the last one is called in the
```

Comment:
I think here we could add links to _Hydrodynamic Solver_ and _ Tracers Solver_  workflows. What do you think?

Reply:
![image](https://user-images.githubusercontent.com/5083518/67429578-d8b36800-f5b6-11e9-93f9-6dabf37055d2.png)

Reply:
@ggrbill 
Do you agree?

Reply:
Yes. Nice!